src/sensors/d200.hpp

```diff
@@ -0,0 +1,157 @@
+const int D200_CMD_PACKET_LEN = 8;
+
+/// @brief hard-coded start command
+const uint8_t D200_START_CMD[] = { 0x54, 0xa0, 0x04, 0, 0, 0, 0, 0x5e };
+
+/// @brief hard-coded stop command
+const uint8_t D200_STOP_CMD[] = { 0x54, 0xa1, 0x04, 0, 0, 0, 0, 0x4a };
+
+/// @brief default scanning speed (rad/s)
+const float D200_DEFAULT_SPEED = (float)(6 * 360) * M_PI / 180.0;
+
+/// @brief min specified scanning speed (rad/s)
+const float D200_MIN_SPEED = (float)(2 * 360 + 1) * M_PI / 180.0;
+
+/// @brief max specified scanning speed (rad/s)
+const float D200_MAX_SPEED = (float)(8 * 360 - 1) * M_PI / 180.0;
+
+struct LidarDataPacket {
+  /// @brief speed of lidar module (rad/s)
+  float lidar_speed = 0.0;
+  
+  /// @brief start angle of measurements (rad)
+  float start_angle = 0.0;
+  
+  /// @brief array of point measurements
+  struct {
+    /// @brief distance (m)
+    float distance = 0.0;
+    /// @brief intensity of measurement. units are ambiguous (not documented)
+    uint8_t intensity = 0;
+  } points[D200_POINTS_PER_PACKET];
+  
+  /// @brief end angle of measurements (rad)
+  float end_angle = 0.0;
+  
+  /// @brief timestamp of measurements, wraps after 30s (s)
+  float timestamp = 0.0;
+};
+
+/// @brief class for LiDAR driver
+class D200LD14P {
+  private:
+    /// @brief default scanning speed (deg/s) (used internally)
+    const uint16_t DEFAULT_SPEED = 6 * 360;
+
+    /// @brief minimum specified scanning speed (deg/s) (used internally)
+    const uint16_t MIN_SPEED = 2 * 360 + 1;
+
+    /// @brief maximum specified scanning speed (deg/s) (used internally)
+    const uint16_t MAX_SPEED = 8 * 360 - 1;
+
+    /// @brief teensy-side cache of packets for comms to use. number of cached packets may be adjusted as needed
+    LidarDataPacket packets[D200_NUM_PACKETS_CACHED];
+
+    /// @brief index of current packet (wraps)
+    int current_packet = 0;
+
+    /// @brief compute CRC8 checksum for buffer
+    /// @param buf pointer to buffer
+    /// @param len length of buffer
+    /// @return CRC8 checksum for buffer
+    uint8_t calc_checksum(uint8_t *buf, int len);
+
+  public:
+    /// @brief constructor and initialization
+    D200LD14P();
+
+    /// @brief set the speed of the LiDAR
+    /// @param speed desired speed of LiDAR (rad/s)
+    void set_speed(float speed);
+
+    /// @brief start the LiDAR motor
+    void start_motor();
+
+    /// @brief stop the LiDAR motor
+    void stop_motor();
+
+    /// @brief read latest packet(s) from D200 module
+    void read();
+
+    /// @brief get the packet array
+    /// @return pointer to start of packet array
+    LidarDataPacket *get_packets() { return packets; }
+
+    /// @brief get the index of the most recent packet
+    /// @return index of latest packet
+    int get_latest_packet_index() { return current_packet; }
+
+    /// @brief get the latest packet
+    /// @return the latest packet
+    LidarDataPacket get_latest_packet() { return packets[current_packet]; }
```

Comment:
What happens if there are no packets ready? Does it return the "previous" latest packet? This behavior should be documented. 

(Currently it looks like it'll crash if the array is empty. Could the array be empty during normal execution (if packet array is cleared)? What about at the very start of execution?).

Reply:
won't crash, the packets are zero initialized but I agree the behavior can be better documented.

Reply:
won't crash, the packets are zero initialized but I agree the behavior can be better documented.